indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionConfigBasedJobTemplate.java

```diff
@@ -70,12 +73,32 @@ public List<CompactionJob> createCompactionJobs(
 
     final List<CompactionJob> jobs = new ArrayList<>();
 
+    CompactionState compactionState = CompactSegments.createCompactionStateFromConfig(config);
+
+    String compactionStateFingerprint = CompactionState.generateCompactionStateFingerprint(
+        compactionState,
+        config.getDataSource()
+    );
+
+    if (segmentIterator.hasNext()) {
```

Comment:
comment here may be useful. Clear up that we pre-write the compaction state we will be creating compaction tasks for so the underlying task does not need to worry about it, all it has to do is persist the fingerprint when it generates segments.